lib/ecto/adapters/postgres/connection.ex

```diff
@@ -178,60 +177,62 @@ if Code.ensure_loaded?(Postgrex) do
+      [" ON CONFLICT ", conflict_target(targets), "DO ", update_all(query, "UPDATE SET ")]
     end
+
     defp conflict_target([]) do
       ""
     end
     defp conflict_target(targets) do
-      "(" <> Enum.map_join(targets, ",", &quote_name/1) <> ") "
+      [?(, intersperse(targets, ?,, &quote_name/1), ?), ?\s]
     end
 
-    defp insert_all([row|rows], counter, acc) do
-      {counter, row} = insert_each(row, counter, "")
-      insert_all(rows, counter, acc <> ",(" <> row <> ")")
-    end
-    defp insert_all([], _counter, "," <> acc) do
-      acc
+    defp insert_all(rows, counter) do
+      intersperse_reduce(rows, ?,, counter, fn row, counter ->
+        {row, counter} = insert_each(row, counter)
+        {[?(, row, ?)], counter}
+      end)
+      |> elem(0)
     end
 
-    defp insert_each([nil|t], counter, acc),
-      do: insert_each(t, counter, acc <> ",DEFAULT")
-    defp insert_each([_|t], counter, acc),
-      do: insert_each(t, counter + 1, acc <> ",$" <> Integer.to_string(counter))
-    defp insert_each([], counter, "," <> acc),
-      do: {counter, acc}
+    defp insert_each(values, counter) do
+      intersperse_reduce(values, ?,, counter, fn
+        nil, counter ->
+          {"DEFAULT", counter}
+        _, counter ->
+          {[?$, Integer.to_string(counter)], counter + 1}
+      end)
+    end
 
     def update(prefix, table, fields, filters, returning) do
-      {fields, count} = Enum.map_reduce fields, 1, fn field, acc ->
-        {"#{quote_name(field)} = $#{acc}", acc + 1}
-      end
+      {fields, count} = intersperse_reduce(fields, ", ", 1, fn field, acc ->
+        {[quote_name(field), " = $", Integer.to_string(acc)], acc + 1}
+      end)
 
-      {filters, _count} = Enum.map_reduce filters, count, fn field, acc ->
-        {"#{quote_name(field)} = $#{acc}", acc + 1}
-      end
+      {filters, _count} = intersperse_reduce(filters, " AND ", count, fn field, acc ->
+        {[quote_name(field), " = $", Integer.to_string(acc)], acc + 1}
+      end)
 
-      assemble(["UPDATE #{quote_table(prefix, table)} SET " <> Enum.join(fields, ", "),
-                "WHERE " <> Enum.join(filters, " AND "),
-                returning(returning)])
+      IO.iodata_to_binary(["UPDATE ", quote_table(prefix, table), " SET ",
+       fields, " WHERE ", filters, returning(returning)])
     end
 
     def delete(prefix, table, filters, returning) do
-      {filters, _} = Enum.map_reduce filters, 1, fn field, acc ->
-        {"#{quote_name(field)} = $#{acc}", acc + 1}
-      end
+      {filters, _} = intersperse_reduce(filters, " AND ", 1, fn field, acc ->
+        {[quote_name(field), " = $", Integer.to_string(acc)], acc + 1}
+      end)
 
-      assemble(["DELETE FROM #{quote_table(prefix, table)} WHERE " <> Enum.join(filters, " AND "),
-                returning(returning)])
+      IO.iodata_to_binary(["DELETE FROM ", quote_table(prefix, table), " WHERE ",
+       filters, returning(returning)])
```

Comment:
Can we please aligning this under the `"` as before? This and the IO.iodata_to_binary calls above. :)